prime-router/docs/design/design/auth/auth-design.md

```diff
@@ -0,0 +1,185 @@
+# Authentication and Authorization design
+
+## Current design
+
+We currently use a variety of methods to authenticate users and authorize actions in our system. 
+We use a mix/match of the following for different endpoints:
+
+- Okta
+  - A cloud-hosted identity and access management service
+- Server2Server
+  - Homegrown JWT-based authn/authz
+- Azure Function Keys
+  - Azure configured shared secrets for authorization
+  - Does not handle authentication at all! (Any user can use any key and the system will work)
+
+## Reasons for updating
+
+- Unify authn/authz to one consistent method as having multiple methods makes it difficult to determine who is using which method
+- Reduce attack surface area since currently we have the weaknesses of each method
+- Minimize authn/authz issues being the cause for missing sent reports
+- Make life easier for Engagement team by having a single choice when onboarding a new sender/receiver
+
+## Requirements
+
+- Protect certain endpoints from only being accessed by those we trust
+- Centralize user management
+- Revoke access if a secret is compromised
+- Audit which clients are accessing which resources
+- Fine-tune access control for specific clients
+- Unify authn/authz approach for all microservices
+- Allow clients the ability to self-service to ease client onboarding
+
+## Identity Provider Options
+
+### Azure AD
+
+Pros
+- Integrated into the Microsoft ecosystem that we currently reside
+- Scales well
+
+Cons
+- Complex setup
+ 
+### Okta
+
+Pros
+- A trusted industry standard
+- Highly scalable
+- Offers self-service to ease Engagement teams workload
+- We already have many of our clients set up here!
+
+Cons
+- Single point of failure if there is an outage
+
+### Homegrown
+
+Pros
+- As flexible as we need it to be
+- Cheap
+
+Cons
+- Custom authn/authz code is often vulnerable to attacks we may not have considered
+- Additional work for us to write and maintain
+
+## Protocol
+
+We will be using OAuth 2.0 with OIDC. OAuth 2.0 is an industry-standard authorization protocol and OIDC is
```

Comment:
I'm not sure OIDC is applicable here since it is for authenticating USERS and backend will deal with machine-machine auth that is not tied to a particular user. OIDC is used presently by the front end for user auth for website, however, if that's what you meant.

Reply:
I believe it is still technically OIDC but perhaps I should expand on this in the doc.

My thought is using the [client credentials flow](https://developer.okta.com/docs/concepts/oauth-openid/#client-credentials-flow) which seems like the right approach.